### ISSUE TITLE
rina-tgen with CBR does weird things when specifying a rate

### ISSUE THREAD

**Issue by edugrasa** [IRATI/traffic-generator#21]
Tried rina-tgen with CBR and rate limited to 1000 kbps (usual normal over shim, 3 IPCPs in the normal), it doesn't generate a constant bit rate.

```
19955(1442965252)#traffic-generator (INFO): Starting test from client traffic.generator.client-1 on port-id 3
19955(1442965252)#traffic-generator (INFO): Duration: 30 s, count: 0 sdus, sdu size: 1400 bytes, reporting interval: 1000 ms
19955(1442965253)#traffic-generator (INFO): Port    3:        93 SDUs (      130200 bytes) in      1008367 us =>   92.2283 p/s,    1.0330 Mb/s
19955(1442965254)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1005168 us =>   88.5424 p/s,    0.9917 Mb/s
19955(1442965255)#traffic-generator (INFO): Port    3:       235 SDUs (      329000 bytes) in      1441345 us =>  163.0422 p/s,    1.8261 Mb/s
19955(1442965256)#traffic-generator (INFO): Port    3:       222 SDUs (      310800 bytes) in      1389317 us =>  159.7907 p/s,    1.7897 Mb/s
19955(1442965257)#traffic-generator (INFO): Port    3:       143 SDUs (      200200 bytes) in      1041364 us =>  137.3199 p/s,    1.5380 Mb/s
19955(1442965259)#traffic-generator (INFO): Port    3:       409 SDUs (      572600 bytes) in      1641326 us =>  249.1888 p/s,    2.7909 Mb/s
19955(1442965259)#traffic-generator (INFO): Port    3:       197 SDUs (      275800 bytes) in      1293343 us =>  152.3184 p/s,    1.7060 Mb/s
19955(1442965261)#traffic-generator (INFO): Port    3:       211 SDUs (      295400 bytes) in      1545350 us =>  136.5386 p/s,    1.5292 Mb/s
19955(1442965262)#traffic-generator (INFO): Port    3:       276 SDUs (      386400 bytes) in      1541349 us =>  179.0639 p/s,    2.0055 Mb/s
19955(1442965262)#traffic-generator (INFO): Port    3:       145 SDUs (      203000 bytes) in      1193363 us =>  121.5054 p/s,    1.3609 Mb/s
19955(1442965263)#traffic-generator (INFO): Port    3:       279 SDUs (      390600 bytes) in      1145348 us =>  243.5941 p/s,    2.7283 Mb/s
19955(1442965264)#traffic-generator (INFO): Port    3:       201 SDUs (      281400 bytes) in      1397324 us =>  143.8464 p/s,    1.6111 Mb/s
19955(1442965265)#traffic-generator (INFO): Port    3:       146 SDUs (      204400 bytes) in      1049358 us =>  139.1327 p/s,    1.5583 Mb/s
19955(1442965267)#traffic-generator (INFO): Port    3:       328 SDUs (      459200 bytes) in      1645366 us =>  199.3477 p/s,    2.2327 Mb/s
19955(1442965267)#traffic-generator (INFO): Port    3:        36 SDUs (       50400 bytes) in      1245322 us =>   28.9082 p/s,    0.3238 Mb/s
19955(1442965269)#traffic-generator (INFO): Port    3:       386 SDUs (      540400 bytes) in      1597338 us =>  241.6520 p/s,    2.7065 Mb/s
19955(1442965270)#traffic-generator (INFO): Port    3:       193 SDUs (      270200 bytes) in      1549304 us =>  124.5721 p/s,    1.3952 Mb/s
19955(1442965270)#traffic-generator (INFO): Port    3:        32 SDUs (       44800 bytes) in      1149348 us =>   27.8419 p/s,    0.3118 Mb/s
19955(1442965271)#traffic-generator (INFO): Port    3:       347 SDUs (      485800 bytes) in      1449369 us =>  239.4145 p/s,    2.6814 Mb/s
19955(1442965272)#traffic-generator (INFO): Port    3:       161 SDUs (      225400 bytes) in      1097349 us =>  146.7172 p/s,    1.6432 Mb/s
19955(1442965273)#traffic-generator (INFO): Port    3:       188 SDUs (      263200 bytes) in      1000088 us =>  187.9835 p/s,    2.1054 Mb/s
19955(1442965274)#traffic-generator (INFO): Port    3:       278 SDUs (      389200 bytes) in      1297358 us =>  214.2816 p/s,    2.4000 Mb/s
19955(1442965276)#traffic-generator (INFO): Port    3:       363 SDUs (      508200 bytes) in      1549357 us =>  234.2907 p/s,    2.6241 Mb/s
19955(1442965276)#traffic-generator (INFO): Port    3:       133 SDUs (      186200 bytes) in      1197337 us =>  111.0798 p/s,    1.2441 Mb/s
19955(1442965277)#traffic-generator (INFO): Port    3:       195 SDUs (      273000 bytes) in      1449300 us =>  134.5477 p/s,    1.5069 Mb/s
19955(1442965278)#traffic-generator (INFO): Port    3:       140 SDUs (      196000 bytes) in      1101332 us =>  127.1188 p/s,    1.4237 Mb/s
19955(1442965279)#traffic-generator (INFO): Port    3:       288 SDUs (      403200 bytes) in      1405327 us =>  204.9345 p/s,    2.2953 Mb/s
19955(1442965280)#traffic-generator (INFO): Port    3:       151 SDUs (      211400 bytes) in      1105354 us =>  136.6078 p/s,    1.5300 Mb/s
19955(1442965281)#traffic-generator (INFO): Port    3:       249 SDUs (      348600 bytes) in      1357363 us =>  183.4439 p/s,    2.0546 Mb/s
19955(1442965282)#traffic-generator (INFO): Port    3:       232 SDUs (      324800 bytes) in      1309355 us =>  177.1865 p/s,    1.9845 Mb/s
19955(1442965282)#traffic-generator (INFO): Port    3:      6346 SDUs,      8884400 bytes in     30309 ms, 2.3450 Mb/s
```

Poisson distribution seems to be working though:

```
19955(1442965174)#traffic-generator (INFO): Starting test from client traffic.generator.client-1 on port-id 3
19955(1442965174)#traffic-generator (INFO): Duration: 30 s, count: 0 sdus, sdu size: 1400 bytes, reporting interval: 1000 ms
19955(1442965175)#traffic-generator (INFO): Port    3:        92 SDUs (      128800 bytes) in      1005106 us =>   91.5326 p/s,    1.0252 Mb/s
19955(1442965176)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1002618 us =>   88.7676 p/s,    0.9942 Mb/s
19955(1442965177)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1001831 us =>   88.8373 p/s,    0.9950 Mb/s
19955(1442965178)#traffic-generator (INFO): Port    3:        90 SDUs (      126000 bytes) in      1006451 us =>   89.4231 p/s,    1.0015 Mb/s
19955(1442965179)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1002210 us =>   88.8037 p/s,    0.9946 Mb/s
19955(1442965180)#traffic-generator (INFO): Port    3:        90 SDUs (      126000 bytes) in      1005272 us =>   89.5280 p/s,    1.0027 Mb/s
19955(1442965181)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1003215 us =>   88.7148 p/s,    0.9936 Mb/s
19955(1442965182)#traffic-generator (INFO): Port    3:        90 SDUs (      126000 bytes) in      1008014 us =>   89.2845 p/s,    1.0000 Mb/s
19955(1442965183)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1005226 us =>   88.5373 p/s,    0.9916 Mb/s
19955(1442965184)#traffic-generator (INFO): Port    3:        90 SDUs (      126000 bytes) in      1009441 us =>   89.1583 p/s,    0.9986 Mb/s
19955(1442965185)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1002117 us =>   88.8120 p/s,    0.9947 Mb/s
19955(1442965186)#traffic-generator (INFO): Port    3:        90 SDUs (      126000 bytes) in      1005844 us =>   89.4771 p/s,    1.0021 Mb/s
19955(1442965187)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1002411 us =>   88.7859 p/s,    0.9944 Mb/s
19955(1442965188)#traffic-generator (INFO): Port    3:        90 SDUs (      126000 bytes) in      1009249 us =>   89.1752 p/s,    0.9988 Mb/s
19955(1442965189)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1002176 us =>   88.8068 p/s,    0.9946 Mb/s
19955(1442965190)#traffic-generator (INFO): Port    3:        90 SDUs (      126000 bytes) in      1008687 us =>   89.2249 p/s,    0.9993 Mb/s
19955(1442965191)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1008325 us =>   88.2652 p/s,    0.9886 Mb/s
19955(1442965192)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1002424 us =>   88.7848 p/s,    0.9944 Mb/s
19955(1442965193)#traffic-generator (INFO): Port    3:        90 SDUs (      126000 bytes) in      1003837 us =>   89.6560 p/s,    1.0041 Mb/s
19955(1442965194)#traffic-generator (INFO): Port    3:        90 SDUs (      126000 bytes) in      1006850 us =>   89.3877 p/s,    1.0011 Mb/s
19955(1442965195)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1002653 us =>   88.7645 p/s,    0.9942 Mb/s
19955(1442965196)#traffic-generator (INFO): Port    3:        90 SDUs (      126000 bytes) in      1008174 us =>   89.2703 p/s,    0.9998 Mb/s
19955(1442965197)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1005562 us =>   88.5077 p/s,    0.9913 Mb/s
19955(1442965198)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1001344 us =>   88.8805 p/s,    0.9955 Mb/s
19955(1442965199)#traffic-generator (INFO): Port    3:        90 SDUs (      126000 bytes) in      1009795 us =>   89.1270 p/s,    0.9982 Mb/s
19955(1442965200)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1005427 us =>   88.5196 p/s,    0.9914 Mb/s
19955(1442965201)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1008444 us =>   88.2548 p/s,    0.9885 Mb/s
19955(1442965202)#traffic-generator (INFO): Port    3:        89 SDUs (      124600 bytes) in      1004824 us =>   88.5727 p/s,    0.9920 Mb/s
19955(1442965203)#traffic-generator (INFO): Port    3:        90 SDUs (      126000 bytes) in      1011506 us =>   88.9762 p/s,    0.9965 Mb/s
```

**Comment by edugrasa** [comment]
May it have to do with storing the result of this operation
https://github.com/IRATI/traffic-generator/blob/master/src/client.cc#L93
in a long variable? (nanos may not have enough room to store the result?)

**Comment by dstaesse** [comment]
don't think so, that long should be able to store almost 300 years worth of nanoseconds.
is this in a VM or on the wall?

**Comment by edugrasa** [comment]
I experience the problems in XEN VMs, Kewin was seeing it in the VWall I think also (we should ask him to confirm)

**Comment by dstaesse** [comment]
This is really weird. What I don't get is that your server reporting interval is all over the place in CBR mode. If it reports only after 1597338 us, it means it didn't get any packets for more than half a second.
I have never seen such behaviour on the virtual wall. Are you running a single tgen per client?
Do you always have this problem with CBR? Never have this problem with poisson?
Can you this test again and have the server report every 20ms? Can you check the wire with tcpdump? Check with the non-blocking version of tgen?

**Comment by edugrasa** [comment]
Yes, single client and single server, always with CBR, never with Poisson.

I'll test it later with the options you suggest and report the results back here.

**Comment by dstaesse** [comment]
I pushed an update to the branch CBR-clock-fix in my repo, can you test it? I don't have time to test it myself right now, but I think it will solve your issue.

**Comment by edugrasa** [comment]
Works like a charm, it fixed the issue :)

```
24596(1443008344)#traffic-generator (INFO): Starting test from client traffic.generator.client-1 on port-id 6
24596(1443008344)#traffic-generator (INFO): Duration: 30 s, count: 0 sdus, sdu size: 1400 bytes, reporting interval: 1000 ms
24596(1443008345)#traffic-generator (INFO): Port    6:       895 SDUs (     1253000 bytes) in      1000475 us =>  894.5751 p/s,   10.0192 Mb/s
24596(1443008346)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000623 us =>  892.4440 p/s,    9.9954 Mb/s
24596(1443008347)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000775 us =>  892.3085 p/s,    9.9939 Mb/s
24596(1443008348)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000950 us =>  892.1525 p/s,    9.9921 Mb/s
24596(1443008349)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1001103 us =>  892.0161 p/s,    9.9906 Mb/s
24596(1443008350)#traffic-generator (INFO): Port    6:       892 SDUs (     1248800 bytes) in      1000143 us =>  891.8725 p/s,    9.9890 Mb/s
24596(1443008351)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000303 us =>  892.7295 p/s,    9.9986 Mb/s
24596(1443008352)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000467 us =>  892.5832 p/s,    9.9969 Mb/s
24596(1443008353)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000616 us =>  892.4503 p/s,    9.9954 Mb/s
24596(1443008354)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000786 us =>  892.2987 p/s,    9.9937 Mb/s
24596(1443008355)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000958 us =>  892.1453 p/s,    9.9920 Mb/s
24596(1443008356)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1001113 us =>  892.0072 p/s,    9.9905 Mb/s
24596(1443008357)#traffic-generator (INFO): Port    6:       892 SDUs (     1248800 bytes) in      1000146 us =>  891.8698 p/s,    9.9889 Mb/s
24596(1443008358)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000294 us =>  892.7375 p/s,    9.9987 Mb/s
24596(1443008359)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000458 us =>  892.5912 p/s,    9.9970 Mb/s
24596(1443008360)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000627 us =>  892.4404 p/s,    9.9953 Mb/s
24596(1443008361)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000786 us =>  892.2987 p/s,    9.9937 Mb/s
24596(1443008362)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000933 us =>  892.1676 p/s,    9.9923 Mb/s
24596(1443008363)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1001098 us =>  892.0206 p/s,    9.9906 Mb/s
24596(1443008364)#traffic-generator (INFO): Port    6:       892 SDUs (     1248800 bytes) in      1000156 us =>  891.8609 p/s,    9.9888 Mb/s
24596(1443008365)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000290 us =>  892.7411 p/s,    9.9987 Mb/s
24596(1443008366)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000471 us =>  892.5796 p/s,    9.9969 Mb/s
24596(1443008367)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000632 us =>  892.4360 p/s,    9.9953 Mb/s
24596(1443008368)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000793 us =>  892.2924 p/s,    9.9937 Mb/s
24596(1443008369)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000943 us =>  892.1587 p/s,    9.9922 Mb/s
24596(1443008370)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1001103 us =>  892.0161 p/s,    9.9906 Mb/s
24596(1443008371)#traffic-generator (INFO): Port    6:       892 SDUs (     1248800 bytes) in      1000154 us =>  891.8627 p/s,    9.9889 Mb/s
24596(1443008372)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000311 us =>  892.7224 p/s,    9.9985 Mb/s
24596(1443008373)#traffic-generator (INFO): Port    6:       893 SDUs (     1250200 bytes) in      1000459 us =>  892.5903 p/s,    9.9970 Mb/s
24596(1443008374)#traffic-generator (INFO): Flow torn down remotely [port-id = 6]
```

**Comment by dstaesse** [comment]
Yeah, guessed so, the clock in your VM drifts. It is common in VMs as the hypervisor stretches and compresses interrupts as it sees fit. Xen is the worst, KVM has a paravirtualised clock and should not have this issue. I think we briefly discussed this a while back.
I'll add a new --distribution option to tgen in "master", default CBR will use this algorithm, CBRC the more accurate 'catchup' algorithm.

**Comment by dstaesse** [comment]
Solved by #22
